compiler/ballerina-lang/src/main/java/org/wso2/ballerinalang/compiler/semantics/analyzer/Types.java

```diff
@@ -4289,6 +4289,28 @@ private BType getIntersection(IntersectionContext intersectionContext, BType lhs
                 return lhsType;
             }
         }
+
+        // TODO: intersections with readonly types are not handled properly. Here, the logic works as follows.
+        // Say we have an intersection called A & readonly and we have another type called B. As per the current
+        // implementation, we cannot easily find the intersection between (A & readonly) and B. Instead, what we
+        // do here is, first find the intersection between A and B then re-create the immutable type out of it.
+
+        if (Symbols.isFlagOn(lhsType.flags, Flags.READONLY) && lhsType.tag == TypeTags.UNION &&
+                ((BUnionType) lhsType).getIntersectionType().isPresent()) {
+            BIntersectionType intersectionType = ((BUnionType) lhsType).getIntersectionType().get();
+            BType finalType = type;
+            List<BType> types = intersectionType.getConstituentTypes().stream().filter(t -> t.tag != TypeTags.READONLY)
+                    .map(t -> getIntersection(intersectionContext, t, env, finalType, visitedTypes))
+                    .filter(Objects::nonNull)
+                    .collect(Collectors.toList());
+            if (types.size() == 1) {
+                BType bType = types.get(0);
+                return ImmutableTypeCloner.getImmutableType(null, this, bType, env, env.enclPkg.packageID, null,
+                        symTable, anonymousModelHelper, names,
+                        new LinkedHashSet<>());
+            }
+        }
```

Comment:
This should be properly handled. Created an issue https://github.com/ballerina-platform/ballerina-lang/issues/33580